src/write.jl

```diff
@@ -442,85 +442,52 @@ function _write_constraint(
-    if Meta.isexpr(expr, :call, 3)
-        @assert expr.args[1] in (:(<=), :(>=), :(<), :(>), :(==))
-        _write_expression(io, variables, expr.args[2])
-        rhs = expr.args[3]
-        if isone(rhs)
-            println(io, " $(expr.args[1]) true;")
-        else
-            @assert iszero(rhs)
-            println(io, " $(expr.args[1]) false;")
-        end
-    else
-        @assert Meta.isexpr(expr, :comparison, 5)
-        error("Two sided not supported")
-    end
+    _write_expression(io, variables, f)
+    println(io, _sense(s), _rhs(s), ";")
     return
 end
 
-function _write_logical_expression(io::IO, variables, expr::Expr)
-    ops = Dict(:|| => "\\/", :&& => "/\\")
-    op = get(ops, expr.head, nothing)
-    @assert op !== nothing
-    print(io, "(")
-    _write_expression(io, variables, expr.args[1])
-    for i in 2:length(expr.args)
-        print(io, " ", op, " ")
-        _write_expression(io, variables, expr.args[i])
-    end
-    print(io, ")")
-    return
-end
-
-_INFIX_OPS = Dict(
-    :- => "-",
-    :+ => "+",
-    :* => "*",
-    :(<) => "<",
-    :(>) => ">",
-    :(<=) => "<=",
-    :(>=) => ">=",
-    :(=>) => "->",
-    :⊻ => "xor",
-)
-
-_PREFIX_OPS = Dict(:(!) => "not")
-
-function _write_call_expression(io::IO, variables, expr::Expr)
-    op = get(_INFIX_OPS, expr.args[1], nothing)
-    if op !== nothing
+function _write_expression(io::IO, variables, f::MOI.ScalarNonlinearFunction)
+    if haskey(_INFIX_OPS, f.head)
+        op = _INFIX_OPS[f.head]
+        @assert length(f.args) > 1
         print(io, "(")
-        _write_expression(io, variables, expr.args[2])
-        for i in 3:length(expr.args)
+        _write_expression(io, variables, f.args[1])
+        for i in 2:length(f.args)
             print(io, " ", op, " ")
-            _write_expression(io, variables, expr.args[i])
+            _write_expression(io, variables, f.args[i])
         end
         print(io, ")")
-    else
-        op = get(_PREFIX_OPS, expr.args[1], nothing)
-        @assert op !== nothing
+    elseif haskey(_PREFIX_OPS, f.head)
+        op = _PREFIX_OPS[f.head]
         print(io, op, "(")
-        _write_expression(io, variables, expr.args[2])
+        _write_expression(io, variables, f.args[1])
+        for i in 2:length(f.args)
+            print(io, ", ")
+            _write_expression(io, variables, f.args[i])
+        end
         print(io, ")")
+    else
+        throw(MOI.UnsupportedNonlinearOperator(f.head))
     end
     return
 end
 
-function _write_expression(io::IO, variables, expr::Expr)
-    if Meta.isexpr(expr, :ref)
-        @assert expr.args[1] == :x
-        _write_expression(io, variables, expr.args[2])
-        return
-    end
-    if Meta.isexpr(expr, :||) || Meta.isexpr(expr, :&&)
-        _write_logical_expression(io, variables, expr)
-    else
-        @assert Meta.isexpr(expr, :call)
-        _write_call_expression(io, variables, expr)
+function _write_expression(io::IO, variables, x::Vector)
```

Comment:
@odow I added this new method to allow writing vector arguments, e.g. for `count([x1, x2, x3])`

Reply:
Ditto here. Why not `count(x1, x2, x3)`?